src/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,253 @@
+    confirmButton: customPropTypes.itemShorthand,
+    defaultOpen: PropTypes.bool,
+    header: customPropTypes.itemShorthand,
+    onCancel: PropTypes.func,
+    onConfirm: PropTypes.func,
+    onOpen: PropTypes.func,
+    open: PropTypes.bool,
+    overlay: customPropTypes.itemShorthand,
+    trapFocus: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
+    trigger: PropTypes.any,
+  }
+
+  static defaultProps = {
+    accessibility: dialogBehavior,
+    actions: {},
+    overlay: {},
+    trapFocus: true,
+  }
+
+  static autoControlledProps = ['open']
+
+  actionHandlers: AccessibilityActionHandlers = {
+    closeAndFocusTrigger: e => {
+      this.handleDialogCancel(e)
+      e.stopPropagation()
+
+      _.invoke(this.triggerRef, 'current.focus')
+    },
+    close: e => this.handleDialogCancel(e),
+  }
+  contentRef = React.createRef<HTMLElement>()
+  triggerRef = React.createRef<HTMLElement>()
+
+  getInitialAutoControlledState(): DialogState {
+    return { open: false }
+  }
+
+  handleDialogCancel = (e: Event | React.SyntheticEvent) => {
+    _.invoke(this.props, 'onCancel', e, { ...this.props, open: false })
+    this.trySetState({ open: false })
+  }
+
+  handleDialogConfirm = (e: React.SyntheticEvent) => {
+    _.invoke(this.props, 'onConfirm', e, { ...this.props, open: false })
+    this.trySetState({ open: false })
+  }
+
+  handleDialogOpen = (e: React.SyntheticEvent) => {
+    _.invoke(this.props, 'onOpen', e, { ...this.props, open: true })
+    this.trySetState({ open: true })
+  }
+
+  handleCancelButtonOverrides = (predefinedProps: ButtonProps) => ({
+    onClick: (e: React.SyntheticEvent, buttonProps: ButtonProps) => {
+      _.invoke(predefinedProps, 'onClick', e, buttonProps)
+      this.handleDialogCancel(e)
+    },
+  })
+
+  handleConfirmButtonOverrides = (predefinedProps: ButtonProps) => ({
+    onClick: (e: React.SyntheticEvent, buttonProps: ButtonProps) => {
+      _.invoke(predefinedProps, 'onClick', e, buttonProps)
+      this.handleDialogConfirm(e)
+    },
+  })
+
+  handleOverlayOverrides = (content: JSX.Element) => (predefinedProps: BoxProps) => ({
+    content,
+    onClick: (e: React.SyntheticEvent, overlayProps: BoxProps) => {
+      _.invoke(predefinedProps, 'onClick', e, overlayProps)
+
+      if (!doesNodeContainClick(this.contentRef.current, e)) {
+        this.handleDialogCancel(e)
+      }
+    },
+  })
+
+  renderComponent({ accessibility, classes, ElementType, styles, unhandledProps }) {
+    const {
+      actions,
+      confirmButton,
+      cancelButton,
+      content,
+      header,
+      overlay,
+      trapFocus,
+      trigger,
+    } = this.props
+    const { open } = this.state
+
+    const dialogContent = (
+      <Ref innerRef={this.contentRef}>
+        <ElementType
+          className={classes.root}
+          {...accessibility.attributes.popup}
+          {...accessibility.keyHandlers.popup}
+          {...unhandledProps}
+        >
+          {Box.create(header, {
```

Comment:
Can we use the `Header` component here? By default we should render H2: https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html

Also, would it be possible to allow adding aria-labelledby (which needs an ID reference to the header)?
The logic should be:
- by default, generate aria-labelledby
- if aria-labelledby provided by the user, use that instead
- if aria-label provided, do not generate aria-labelledby

For aria-describedby I would leave it fully on the consumer to provide correct referecne as not all dialogs have clear description.

Reply:
Replaced with `Header`, https://github.com/stardust-ui/react/pull/790/commits/fda25b76837c3d35bc3c45b2c15d626f9d95b8ee

Reply:
Created an issue for `aria`s, #812